aip/0180.md

```diff
@@ -0,0 +1,126 @@
+APIs that is then launched into a production service with the expectation that
+it continues to work (unless the API has a [stability level][aip-181] that
+indicates otherwise). Therefore, it is important to understand what constitutes
+a backwards compatible change and what constitutes a backwards incompatible
+change.
+
+## Guidance
+
+Existing client code **must not** be broken by a service updating to a new
+minor or patch release. Old clients **must** be able to work against newer
+servers (with the same major version number).
+
+**Important:** It is not always salient whether a change is compatible or not.
+The guidance here **should** be treated as indicative, rather than as a
+comprehensive list of every possible change.
+
+There are three distinct types of compatibility to consider:
+
+1. Source compatibility: Code written against a previous version **must**
+   compile against a newer version, and successfully run with a newer version
+   of the client library.
+2. Wire compatibility: Code written against a previous version **must** be able
+   to communicate correctly with a newer server. In other words, not only are
+   inputs and outputs compatible, but the serialization and deserialization
+   expectations continue to match.
+3. Semantic compatibility: Code written against a previous version **must**
+   continue to receive the intended and expected results.
+
+### Adding components
+
+In general, new components (interfaces, methods, messages, fields, enums, or
+enum values) **may** be added to existing APIs in the same major version.
+
+However, keep the following guidelines in mind when doing this:
+
+- Code written against the previous surface (and thus is unaware of the new
+  components) **must** continue to be treated the same way as before.
+  - New required fields **must not** be added to existing request messages or
+    resources.
+  - Any field being populated by clients **must** have a default behavior
+    matching the behavior before the field was introduced.
+  - Any field previously populated by the server **must** continue to be
+    populated, even if it introduces redundancy.
+- Enums **may** be freely added to request messages. For response messages and
+  resources, be aware that it is possible that user code does not handle new
+  values gracefully. In this situation, enums which are expected to receive new
+  values **should** document this, and while enum values still **may** be added
+  in this situation, appropriate caution **should** be used.
+
+**Note:** It is possible when adding a component closely related to an existing
+component (for example, `string foo_value` when `string foo` already exists) to
+enter a situation where generated code will conflict. Service owners **should**
+be aware of subtleties in the tooling they or their users are likely to use
+(and tool authors **should** endeavor to avoid such subtleties if possible).
```

Comment:
Hmm. That's a pretty tall ask, given that there could be proto generators for languages I've never used.
What are the compatibility rules for the Rust generator, for example? I've no idea. Usually it's up to the proto generator to avoid this sort of thing - for example, if someone has a field called `types`, the C# generator will generate `Types_` as the property name to avoid `Types` which is used for nested types.

Reply:
I agree, but I do not really see any alternative. In hindsight, I think that tooling (like protobuf) should have been a _lot_ more hesitant to do this sort of thing than they have been, but in the real world we live in, these conflicts happen, meaning API owners that want to avoid them may need to gain some awareness.

One thing that could help with this is compiling some kind of "known potential conflicts" wiki page or something similar, but that is out of scope for this AIP.

Reply:
Discussed in person and @jskeet and I agree that there is no real way around this.